cmake/FlatccGenerateSources.cmake

```diff
@@ -0,0 +1,329 @@
+# Use `flatcc_generate_sources` to generate C source files from flatbuffer definition files.
+# The sources will be generated at build time, and only if some target has the output target
+# as a dependency or adds the generated sources.
+#
+#   flatcc_generate_sources(
+#      <name>
+#      DEFINITION_FILES <flatbuffer-definition-file> [<flatbuffer-definition-file> [...]]
+#      [OUTPUT_DIR <output-directory>]
+#      [ALL] [SCHEMA] [COMMON] [COMMON_READER] [COMMON_BUILDER] [BUILDER] [READER]
+#      [VERIFIER] [JSON_PARSER] [JSON_PRINTER] [JSON] [RECURSIVE]
+#      [OUTFILE <output-file>] [PREFIX <prefix>]
+#      [PATHS <include-path> [<include-path> [...]]]
+#      [COMPILE_FLAGS <flatcc-flag> [<flatcc-flag> [...]]]
+#   )
+#
+# The following things are returned:
+#
+#   <name>_GENERATED_SOURCES
+#
+#     Variable containing a list of all generated sources.
+#     This variable can be added to `add_library`/`add_executable`.
+#
+#   flatcc_generated::<name>
+#
+#     Target that contains all headers. Users must link to it.
+#     When linking to it, the headers will be generated.
+#     The folder containing these headers will also be added to the include path.
+#
+# Required arguments:
+#
+#   <name>
+#
+#     Unique name. It is used for output variable(s) and target(s).
+#
+#   DEFINITION_FILES <definition-file> [<definition-file> [...]]
+#
+#     Flatcc will generate sources for all definition files listed here.
+#     This function also track included files, so it is not needed to list all dependent files here.
+#     Because the search for included files happens at configure time,
+#     the definition files must be available before calling this function.
+#
+# Optional arguments:
+#
+#   OUTPUT_DIR <output-directory>
+#
+#     All generated sources will be written in the `output-directory` folder.
+#
+#   ALL
+#   SCHEMA
+#   COMMON
+#   COMMON_READER
+#   COMMON_BUILDER
+#   BUILDER
+#   READER
+#   VERIFIER
+#   JSON_PARSER
+#   JSON_PRINTER
+#   JSON
+#   RECURSIVE
+#
+#     When specified, these will instruct flatcc to generate a specific type of source code.
+#     It is preferable to use these options instead of adding flatcc arguments to COMPILE_DEFINITIONS.
+#     This is because the generated sources will have correct dependency information.
+#     When none of these are specified, the default is READER.
+#     These options can be combined.
+#
+#   OUTFILE <output-file>
+#
+#     Write all source into one file named `output-file`.
+#
+#   PREFIX <prefix>
+#
+#     Prefix all symbols with `prefix`
+#
+#   TARGET <target>
+#
+#     Create a custom target named `target` that will generate the sources.
+#
+#   PATHS <include-path> [<include-path> [...]]
+#
+#     Add extra include search paths where flatcc should look for included defintion files.
+#
+#   COMPILE_FLAGS <flatcc-flag> [<flatcc-flag> [...]]
+#
+#     Add extra arguments to flatcc.
+#
+
+cmake_minimum_required(VERSION 3.3)
```

Comment:
I had to bump the minimum required cmake version to 3.3, to enable this usage on MSVC:
```
include(FlatccGenerateSources)
flatcc_generate_sources(flatbuffer_headers DEFINITION_FILES input.fbs ALL)

add_executable(main main.cc)
target_link_libraries(main PRIVATE flatcc::runtime flatcc_generated::flatbuffer_headers)
```

Reply:
We should keep an example without flatcc_generate_sources and make a README comment about this in build section and about disabling tests for older CMake versions.

Reply:
I have kept `test/monster_test_solo` as a `add_custom_command`, because it was using `--stdout`, which is not used by the `flatcc_generate_sources` function.
Is that enough?

Reply:
Yes, but I suggest adding a note in the README.